contracts/Tokens/VTokens/VToken.sol

```diff
@@ -1365,80 +1365,204 @@ contract VToken is VTokenInterface, Exponential, TokenErrorReporter {
-                    ((stableBorrows * averageStableBorrowRate) - (stableDebt * stableRateMantissa)) /
-                    stableBorrowsNew;
-            }
+        uint256 accountBorrowsNew = stableDebt + amount;
+        uint256 stableBorrowsNew;
+        uint256 averageStableBorrowRateNew;
 
-            /////////////////////////
-            // EFFECTS & INTERACTIONS
-            // (No safe failures beyond this point)
-            accountBorrows[account].principal = accountBorrowsNew;
-            accountBorrows[account].interestIndex = borrowIndex;
+        if (InterestRateMode(rateMode) == InterestRateMode.STABLE) {
+            require(variableDebt > 0, "vToken: swapBorrowRateMode variable debt is 0");
+            require(variableDebt >= amount, "Insufficient amount");
+
+            (stableBorrowsNew, averageStableBorrowRateNew) = _updateStatesForStableRateSwap(
+                amount,
+                stableDebt,
+                variableDebt,
+                account,
+                accountBorrowsNew
+            );
+        } else {
+            require(stableDebt > 0, "vToken: swapBorrowRateMode stable debt is 0");
+            require(stableDebt >= amount, "Insufficient amount");
 
-            accountStableBorrows[account].principal = 0;
-            accountStableBorrows[account].interestIndex = stableBorrowIndex;
+            (stableBorrowsNew, averageStableBorrowRateNew) = _updateStatesForVariableRateSwap(
+                amount,
+                stableDebt,
+                variableDebt,
+                account
+            );
         }
 
         stableBorrows = stableBorrowsNew;
         averageStableBorrowRate = averageStableBorrowRateNew;
 
-        emit SwapBorrowRateMode(account, rateMode);
+        emit SwapBorrowRateModeWithAmount(account, rateMode, amount);
+    }
+
+    /**
+     * @notice Rebalances the stable interest rate of a user to the current stable borrow rate.
+     * - Users can be rebalanced if the following conditions are satisfied:
+     *     1. Utilization rate is above rebalanceUtilizationRateThreshold.
+     *     2. Average market borrow rate should be less than the rebalanceRateFractionThreshold fraction of variable borrow rate.
+     * @param account The address of the account to be rebalanced
+     * custom:events RebalancedStableBorrowRate - Emits after rebalancing the stable borrow rate for the user.
+     **/
+    function rebalanceStableBorrowRate(address account) external {
+        accrueInterest();
+
+        validateRebalanceStableBorrowRate();
+        _updateUserStableBorrowBalance(account);
+
+        uint256 stableBorrowRate = stableBorrowRatePerBlock();
+
+        accountStableBorrows[account].stableRateMantissa = stableBorrowRate;
+
+        emit RebalancedStableBorrowRate(account, stableBorrowRate);
+    }
+
+    /// Validate the conditions to rebalance the stable borrow rate.
+    function validateRebalanceStableBorrowRate() public view {
```

Comment:
Does it make sense for this view to return a value or is it sufficient that the caller assumes success if it doesn't revert?

Reply:
Fixed, returning true if all conditions get satisfied.